### PR TITLE
Feat/job application table

### LINKED ISSUE #15: Create Job Applications table
## Dependancies
- [x] Companies table must be created

## Acceptance Criteria
- [x] Create a new migration to add a "job application" table to the database
- [x] seed the database
- [x] add CRUD functionality to create that job application.
- [x] create route
- [x] return 200 status code upon successful creation
- [x] happy path testing
- [x] sad path testing
- [x] make companies a foreign key on job_application table

### PR DESCRIPTION
**Pull request by jchirch**
## Type of Change
- [x] feature ⛲
- [x] documentation update 📃
- [x] testing 🧪

## Description
This PR adds the JobDescription table to the database and incorporates migrations to add associations to other tables. Changes route path for users to access job requests specific to user id.

## Motivation and Context
This adds functionality to the app by allowing users to create and save prospective jobs to the database. Incorporates testing, including sad paths and edge cases.
Closes #15 

## Added Test?
- [x] Yes 🫡
  Model testing for associations and validations.
  Requests for post network requests.
- [x] All previous tests still pass 🥳

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: config/routes.rb
Comment (MDelarosa1993) [inline config/routes.rb:1]: Would the endpoint for the jobs application routes would be "api/v1/users/user.id/jobapplications"?

File: app/controllers/api/v1/job_applications_controller.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+      :notes, 
+      :job_description, 
+      :application_url, 
+      :contact_information, 
+      :company_id
+    )
+  end
+end
+
+# class Api::V1::JobApplicationsController < ApplicationController
```
Comment (kylomite) [inline app/controllers/api/v1/job_applications_controller.rb:31]: I'm assuming this is the code from before we refactored to route through users. Let me know if I can delete these comments.

File: app/models/job_application.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+  validates :position_title, presence: true
+  validates :date_applied, presence: true
+  validates :status, presence: true
+  validates :notes, presence: true
+  validates :job_description, presence: true
+  validates :application_url, presence: true
+  validates :contact_information, presence: true
+  validates :company_id, presence: true
+end
```
Comment (jphill19) [inline app/models/job_application.rb:13]: Nice, this is simple and clean way to start off the Job Applications. there are no validations, but this is something we can raise in a new issue to accomplish

File: app/models/job_application.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+class JobApplication < ApplicationRecord
+  belongs_to :company
+  belongs_to :user
+
+  validates :position_title, presence: true
+  validates :date_applied, presence: true
+  validates :status, presence: true
+  validates :notes, presence: true
+  validates :job_description, presence: true
+  validates :application_url, presence: true
```
Comment (kylomite) [inline app/models/job_application.rb:10]: We talked about this on zoom, I am going to make a new ticket for validating  the uniqueness of job apps, this will prevent a user from saving multiple duplicate instances of job applications.

File: spec/models/job_application_spec.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    it { should validate_presence_of(:application_url) }
+    it { should validate_presence_of(:contact_information) }
+    it { should validate_presence_of(:company_id) }
+  end
+
+  describe "associations" do
+    it { should belong_to(:company) }
+    it { should belong_to(:user) }
+  end
+end
```
Comment (jphill19) [inline spec/models/job_application_spec.rb:19]: I'm glad you created test the Job Applications.

File: spec/requests/api/v1/job_applications/job_application_create_spec.rb [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        expect(response).to have_http_status(:bad_request)
+        expect(response.status).to eq(400)
+
+        json = JSON.parse(response.body, symbolize_names: true)
+
+        expect(json[:message]).to eq("Company must exist and Position title can't be blank")
+        expect(json[:status]).to eq(400)
+      end
+    end
+  end
```
Comment (jphill19) [inline spec/requests/api/v1/job_applications/job_application_create_spec.rb:97]: Nice, these are some nice Sad Path test for the create a job application route.

File: app/models/job_application.rb [hunk 6]
```diff
@@ -0,0 +1,13 @@
+class JobApplication < ApplicationRecord
+  belongs_to :company
+  belongs_to :user
+
+  validates :position_title, presence: true
+  validates :date_applied, presence: true
+  validates :status, presence: true
+  validates :notes, presence: true
```
Comment (kylomite) [inline app/models/job_application.rb:8]: Im not sure if notes should be required for a job application? Not all users will want to post a note in every situation

Comment (jphill19) [inline app/models/job_application.rb:8]: this is one thing that should not be required, in my opinion, as this seems like an optional feature.

File: spec/requests/api/v1/job_applications/job_application_create_spec.rb [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+      @facebook = Company.create!(user_id: @user.id, name: "Facebook", website: "facebook.com", street_address: "1 Hacker Way", city: "Menlo Park", state: "CA", zip_code: "94025", notes: "Social media")
+
+      @amazon = Company.create!(user_id: @user.id, name: "Amazon", website: "amazon.com", street_address: "410 Terry Ave N", city: "Seattle", state: "WA", zip_code: "98109", notes: "E-commerce")
+    end
+    let(:job_application_params) do
+      {
+        position_title: "Jr. CTO",
+        date_applied: "2024-10-31",
+        status: 1,
+        notes: "Fingers crossed!",
```
Comment (kylomite) [inline spec/requests/api/v1/job_applications/job_application_create_spec.rb:19]: If we update the notes attribute to not be required, it would be nice to see examples of job applications without notes for variety.

**Review by kylomite — CHANGES_REQUESTED** [review]
Really good progress on this ticket, I will work on getting a new ticket up for adding uniqueness to application url

File: app/controllers/api/v1/job_applications_controller.rb [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+#       :position_title, 
+#       :date_applied, 
+#       :status, 
+#       :notes, 
+#       :job_description, 
+#       :application_url, 
+#       :contact_information, 
+#       :company_id
+#     )
+#   end
```
Comment (jphill19) [inline app/controllers/api/v1/job_applications_controller.rb:55]: There are some remaining comments that be left out

**Review by jphill19 — COMMENTED** [review]
The code looks great, I would remove the comments and potentially change the validation for notes not to be required

File: app/models/job_application.rb (hunk 6 above)
Comment (jchirch) [reply app/models/job_application.rb:8]: Good catch, in hindsight I agree, not everyone wants to submit notes. I will correct this and resubmit.

File: app/controllers/api/v1/job_applications_controller.rb (hunk 1 above)
Comment (jchirch) [reply app/controllers/api/v1/job_applications_controller.rb:31]: That's an oversight on my part, I will tidy this up.

File: config/routes.rb
Comment (jphill19) [reply config/routes.rb:1]: Correct, it might be best if stick to this design and adhere to RESTful design.

File: app/models/job_application.rb (hunk 6 above)
Comment (jchirch) [reply app/models/job_application.rb:8]: Been receiving a lot of valid feedback on this, I will change and resubmit.

**Review by kylomite — APPROVED** [review]
Thank you for hearing me, and taking my feedback into account, and in such a timely manner. Good job my friend

**Comment by ROlearyPro** [comment]
While I haven't comprehended some of the backend context, I can tell you that Joe was working on this quite hard, and took his time in order to get a solution that everyone would be happy with. Nice work!
